src/sctools/metrics/aggregator.py

```diff
@@ -0,0 +1,319 @@
+        self._genomic_read_quality = OnlineGaussianSufficientStatistic()
+
+        # alignment location information
+        self.reads_mapped_exonic = 0
+        self.reads_mapped_intronic = 0
+        self.reads_mapped_utr = 0
+
+        # todo implement this once we have a gene model
+        # self.reads_mapped_outside_window = 0  # reads should be within 1000 bases of UTR
+        # self._read_distance_from_termination_site = OnlineGaussianSufficientStatistic()
+
+        # alignment uniqueness information
+        self.reads_mapped_uniquely = 0
+        self.reads_mapped_multiple = 0
+        self.duplicate_reads = 0
+
+        # alignment splicing information
+        self.spliced_reads = 0
+        self.antisense_reads = 0
+        self._plus_strand_reads = 0  # strand balance  # todo implement property here
+
+        # higher-order methods, filled in by finalize() when all data is extracted
+        self.molecule_barcode_fraction_bases_above_30_mean: float = None
+        self.molecule_barcode_fraction_bases_above_30_variance: float = None
+        self.genomic_reads_fraction_bases_quality_above_30_mean: float = None
+        self.genomic_reads_fraction_bases_quality_above_30_variance: float = None
+        self.genomic_read_quality_mean: float = None
+        self.genomic_read_quality_variance: float = None
+        self.n_molecules: float = None
+        self.n_fragments: float = None
+        self.reads_per_molecule: float = None
+        self.reads_per_fragment: float = None
+        self.fragments_per_molecule: float = None
+        self.fragments_with_single_read_evidence: int = None
+        self.molecules_with_single_read_evidence: int = None
+
+    def finalize(self):
+        """
+        calculate higher-order metrics which cannot be calculated on-the-fly from individual
+        molecules, but rather, require all information for an instance to be compiled
+        """
+
+        self.molecule_barcode_fraction_bases_above_30_mean: float = \
+            self._molecule_barcode_fraction_bases_above_30.mean
+
+        self.molecule_barcode_fraction_bases_above_30_variance: float = \
+            self._molecule_barcode_fraction_bases_above_30.calculate_variance()
+
+        self.genomic_reads_fraction_bases_quality_above_30_mean: float = \
+            self._genomic_reads_fraction_bases_quality_above_30.mean
+
+        self.genomic_reads_fraction_bases_quality_above_30_variance: float = \
+            self._genomic_reads_fraction_bases_quality_above_30.calculate_variance()
+
+        self.genomic_read_quality_mean: float = self._genomic_read_quality.mean
+
+        self.genomic_read_quality_variance: float = self._genomic_read_quality.calculate_variance()
+
+        self.n_molecules: int = len(self._molecule_histogram.keys())
+
+        self.n_fragments: int = len(self._fragment_histogram.keys())
+
+        try:
+            self.reads_per_molecule: float = self.n_reads / self.n_molecules
```

Comment:
Just wondering, when would you have zero molecules?

Reply:
When every read for a cell is unmapped.